clang/test/AST/attr-print-emit.cpp

```diff
@@ -73,3 +73,15 @@ class C {
   // CHECK: void pwtt(void *, int) __attribute__((pointer_with_type_tag(foo, 2, 3)));
   void pwtt(void *, int) __attribute__((pointer_with_type_tag(foo, 2, 3)));
 };
+
+#define ANNOTATE_ATTR __attribute__((annotate("Annotated")))
+ANNOTATE_ATTR int annotated_attr ANNOTATE_ATTR = 0;
+// CHECK: __attribute__((annotate("Annotated"))) int annotated_attr __attribute__((annotate("Annotated"))) = 0;
+
+// FIXME: We do not print the attribute as written after the type specifier.
+int ANNOTATE_ATTR annotated_attr_fixme = 0;
```

Comment:
Note that attributes can appear after the type specifier, we print them on the left but not where they appeared in code unfortunately.

Reply:
Unfortunate, but not a deal breaker IMO.

Reply:
I think we could make this work but I believe it's not worth it at that stage...

Reply:
Hmmm, not certain if it's a deal breaker, but it sure comes close -- we support roughly 70 type attributes, including common ones like calling conventions and address spaces. So this is incremental improvement, but doesn't actually resolve the needs of either @vgvassilev or @giulianobelinassi for being able to print attributes and still compile the code. e.g., `int * [[clang::address_space(0)]] i;`

Reply:
With the following input:
```
int * __attribute__((address_space(0))) i;
int * [[clang::address_space(0)]] j;
```
This prints:
```
int *__attribute__((address_space(0))) i;
int *__attribute__((address_space(0))) j
```
which seems to be semantically the same thing.

Reply:
I would expect it to retain the original spelling of the attribute rather than swap between GNU and [[]] styles (because syntax matters greatly for appertainment with [[]] style attributes but does not [as much] for GNU-style).

Also, how does it handle monstrosities like: https://godbolt.org/z/95Pq7orde (note, implicit int behavior is definitely not a deal breaker, lol)

Reply:
Is that something this PR breaks? I am not sure I understand if I need to change anything.

Reply:
The PR changes behavior from one really broken state to a slightly less but still broken state. Consider:
```
[[clang::annotate("do the")]] int [[clang::annotate_type("test")]] * [[clang::annotate_type("again")]] i [[clang::annotate("and again!")]] = 0;
```
Currently we print this back as: `int *i [[clang::annotate_type(...)]] [[clang::annotate_type(...)]] [[clang::annotate("do the")]] [[clang::annotate("and again!")]] = 0;` which is all kinds of wrong. Now we print this back as: `[[clang::annotate("do the")]] int *i [[clang::annotate_type(...)]] [[clang::annotate_type(...)]] [[clang::annotate("and again!")]] = 0;` which is in better shape but still gets the type attributes wrong by moving the attribute off `int` and `*` and onto the declaration of `i`. The new code won't compile (it tries to add type attributes to a declaration, and the `...` instead of the original string literal won't compile either).

Given that printing is best-effort, perhaps this is fine because it's incremental improvement, but we're still not quite there in terms of accurate printing either.